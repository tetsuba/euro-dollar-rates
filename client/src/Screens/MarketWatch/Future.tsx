import React, {useState, useEffect} from 'react'
import {useQuery} from "react-query";
import {useParams} from "react-router-dom";
import {fetchFuture} from '../../utils/service'

// COMPONENTS
import BreadCrumbs from '../../Components/BreadCrumbs/BreadCrumbs'
import LineChart from '../../Components/Charts/LineChart'
import FutureCheckBoxList from '../../Components/List/FutureCheckBoxList'
import LoadingSpinner from '../../Components/LoadingSpinner/LoadingSpinner'

// UTILS
import {mutateFutureListForLineChart} from '../../utils/mutations'
import { EventClickType } from '../../utils/types'
import Error from "../Error";


export default function Future() {
    const [toggleChart, setToggleChart] = useState(false)
    const [list, setList] = useState([])
    const symbol = useParams()
    const {isLoading, data, isSuccess, isError} = useQuery(['future', symbol], fetchFuture)

    useEffect(() => {
        if(isSuccess){
            // @ts-ignore
            setList(data.list)
        }
    }, [data, isSuccess])

    if (isLoading) return (<LoadingSpinner />)
    if (isError) return <Error />

    return (
        <>
            <BreadCrumbs />
            <>
                <div className="row">
                    <button
                        type="button"
                        className="btn btn-primary"
                        onClick={() => setToggleChart(!toggleChart)}
                    >
                        {toggleChart ? 'Back' : 'Create Chart'}
                    </button>
                </div>
                {
                    !toggleChart && (
                        <div className="row mt-5" data-testid="future-checkbox-list">
                            <FutureCheckBoxList
                                list={list}
                                onClick={
                                    (e: EventClickType): void => {
                                        const el = e.target as HTMLElement
                                        const index: null | string = el.getAttribute('data-index')
                                        const checked: null | string = el.getAttribute('data-checked')

                                        // TODO: Look into this and find out why TS is producing this error
                                        // TS2339: Property 'checked' does not exist on type 'never'.
                                        // @ts-ignore
                                        list[Number(index)].checked = !(checked === 'true')
                                        // Return a new array to update components
                                        setList([...list])
                                    }
                                }
                            />
                        </div>
                    )
                }
                {
                    toggleChart && (<div className="row mt-5"><LineChart {...mutateFutureListForLineChart(list)} /></div>)
                }
            </>
        </>
    )
}
