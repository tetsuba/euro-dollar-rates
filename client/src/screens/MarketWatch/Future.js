import React from 'react'
import { getFuture } from '../../utils/service'
import { withRouter } from '../../HOC/withRouter'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'
import { convertFutureStringToObject } from './utils'

class Future extends React.Component {

    state = {
        future: {}
    }

    async componentDidMount() {
        const future = await getFuture(this.props.params.symbol)

        this.setState({
            future: future.data,
            futureYear: future.data.list.map(convertFutureStringToObject)
        })
    }

    renderList() {
        return this.state.future.list.map(({name, price}) => (
          <div>
              <span>{name}</span>
          </div>))
    }

    render() {
        console.log(this.state)
        return (
          <>
              <BreadCrumbs />
              { this.state.future.list && this.renderList() }
          </>
        )
    }
}

export default withRouter(Future)