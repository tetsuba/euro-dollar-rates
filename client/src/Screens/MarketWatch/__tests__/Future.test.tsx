import React from 'react'
import axios from 'axios'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import Future from '../Future'
import 'jest-canvas-mock'

jest.mock('axios')

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        symbol: 'ED',
    }),
}))

jest.mock('chart.js', () => {
    return jest.fn()
})

const mockData = {
    data: {
        list: [
            {
                year: 2021,
                month: '06',
                price: '1,000,000.50',
                name: 'Eurodollar 3 Month Jun 2021',
                symbol: 'EDM21',
            },
            {
                year: 2010,
                month: '06',
                price: '100,000',
                name: 'Eurodollar 3 Month Jun 2010',
                symbol: 'EDM10',
            },
            {
                year: 1999,
                month: '06',
                price: '1,000',
                name: 'Eurodollar 3 Month Jun 1999',
                symbol: 'EDM99',
            },
            {
                year: 2001,
                month: '06',
                price: '10,000',
                name: 'Eurodollar 3 Month Jun 2000',
                symbol: 'EDM01',
            },
        ],
    },
}

describe('Future', () => {
    // @ts-ignore
    axios.get.mockReturnValueOnce(mockData)

    const rendered = render(
        <MemoryRouter
            initialEntries={['', { pathname: '/market-watch/futures/ED' }]}
        >
            <Future />
        </MemoryRouter>
    )

    test('should render', async () => {
        await waitFor(() => expect(screen.getByTestId('future-checkbox-list')))
        expect(rendered.asFragment()).toMatchSnapshot()

        /* Event: onClick
         * Button: Create Chart
         * Test: should hide future checkbox list and display line chart
         *  */
        fireEvent.click(rendered.getByText(/Create Chart/i))
        await waitFor(() => expect(rendered.getByTestId('line-chart')))
        expect(rendered.queryByTestId('future-checkbox-list')).toBeNull()

        /* Event: onClick
         * Button: Back
         * Test: should hide line chart and display future checkbox list
         *  */
        fireEvent.click(rendered.getByText(/Back/i))
        await waitFor(() =>
            expect(rendered.getByTestId('future-checkbox-list'))
        )
        expect(rendered.queryByTestId('line-chart')).toBeNull()

        /* Test: all checkboxes should be checked = 'true'
         *  */
        rendered.queryAllByTestId('checkbox-custom').forEach((node) => {
            expect(node.getAttribute('data-checked')).toMatch('true')
        })

        /* Event: onClick
         * Checkbox: uncheck
         * Test: should uncheck the first checkbox in the list
         *  */
        const firstCheckBox = rendered.queryAllByTestId('checkbox-custom')[0]
        fireEvent.click(firstCheckBox)
        expect(firstCheckBox.getAttribute('data-checked')).toMatch('false')
    })
})
