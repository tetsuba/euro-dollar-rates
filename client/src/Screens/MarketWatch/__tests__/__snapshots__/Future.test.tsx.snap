// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Future should render 1`] = `
<DocumentFragment>
  <div
    class="row shadow mb-5"
  >
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb mt-3"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumb-item"
        >
          <a
            href="/market-watch"
          >
            market-watch
          </a>
        </li>
        <li
          class="breadcrumb-item"
        >
          <a
            href="/market-watch/futures"
          >
            futures
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        >
          ED
        </li>
      </ol>
    </nav>
  </div>
  <div
    class="row"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      Create Chart
    </button>
  </div>
  <div
    class="row mt-5"
    data-testid="future-checkbox-list"
  >
    <ul
      class="list-group list-group-flush"
    />
  </div>
</DocumentFragment>
`;
